### ISSUE TITLE
Pyright reports reportMissingSuperCall in a class that is not a subclass

### ISSUE THREAD

**Issue by diego-pm** [microsoft/pyright#3968]
**Describe the bug**
Pyright is reporting `reportMissingSuperCall` when defining the `__init__` method of a class that is not subclassing any other class.
This happens when I have the configuration parameter `reportMissingSuperCall` enabled.

**To Reproduce**
Enable the `reportMissingSuperCall` option in the pyright configuration file.
Then, run pyright on the following snippet:
```python
class A:
    def __init__(self) -> None:
        print("Init")
```

**Expected behavior**
No errors should be reported because class `A` is not subclassing any other classes.

**VS Code extension or command-line**
pyright CLI 1.1.271

**Comment by erictraut** [comment]
The feature is working as intended.

The `reportMissingSuperCall` is an extremely strict (and therefore opinionated) check. It isn't even enabled in strict mode and must be explicitly enabled by name if you want to use it. See [this issue](https://github.com/microsoft/pyright/issues/2789) for the discussion and feedback that led to this feature.

In your example, class `A` is deriving from `object`, so a call to `super().__init__()` would invoke `object.__init__`.

**Comment by diego-pm** [comment]
I see, thank you. From reading the [docs](https://github.com/microsoft/pyright/blob/main/docs/configuration.md#:~:text=reportMissingSuperCall%20%5Bboolean%20or%20string%2C%20optional%5D%3A%20Generate%20or%20suppress%20diagnostics%20for%20__init__%2C%20__init_subclass__%2C%20__enter__%20and%20__exit__%20methods%20in%20a%20subclass%20that%20fail%20to%20call%20through%20to%20the%20same%2Dnamed%20method%20on%20a%20base%20class.%20The%20default%20value%20for%20this%20setting%20is%20%27none%27.), I thought that the subclass should be made explicit. I do not understand exactly why it is needed the `__init__()` method for this type of classes, as all classes inherit from `object`.

**Comment by erictraut** [comment]
Yeah, there's no such thing in Python as a "class that is not a subclass". If you provide no base class list in the class definition, the class implicitly inherits from `object`. Python 2 required you to explicitly include `object` as a base class, but this requirement was dropped in Python 3 for convenience. In Python 3, `class A:` is equivalent to `class A(object):`.

**Comment by darlingm** [comment]
@erictraut, in #2789, your original comment is: (added bold/italics, and spaces around init to prevent styling)

> When a parent class _**(other than object)**_ defines an __ init __ method that is not called by a subclass from its own __ init __ method, an error should probably be generated.

Couldn't we have reportMissingSuperCall ignore classes that only have `object` as their parent class to make this feature more useful, less annoying, and follow your original idea on it?

**Comment by erictraut** [comment]
> Couldn't we have reportMissingSuperCall ignore classes that only have object as their parent class to make this feature more useful, less annoying, and follow your original idea on it?

I don't think it's safe to skip a call to `super.__init__` in this case. In particular, if the class were used in a multiple-inheritance class hierarchy, it may still be necessary to call `super.__init__()`. I suppose it would be safe to skip this if the class were marked `@final`, but that's probably not what you're looking for.

**Comment by darlingm** [comment]
> I don't think it's safe to skip a call to `super.__init__` in this case. In particular, if the class were used in a multiple-inheritance class hierarchy, it may still be necessary to call `super.__init__()`. I suppose it would be safe to skip this if the class were marked `@final`, but that's probably not what you're looking for.

Ahh, I see!  I thought of a few ways to maybe have a useful `reportMissingSuperCall` that wouldn't drive some people crazy with it reporting on things we can see are unnecessary.

* An option `reportMissingSuperCallExceptToObject`?  A user who isn't going to use `reportMissingSuperCall` due to this issue could choose the new option to get some coverage (albeit incomplete) rather than nothing.
* A boolean flag `ignoreMissingSuperCallToObject`?  Same reason.
* Have `reportMissingSuperCall` ignore calls to `super.__init__` with `object` if there's no multiple-inheritance hierarchy necessitating it?  I'm not sure if Pyright can easily look through a complete inheritance hierarchy, such as subclasses not mentioned in the source file or any of the file's imports.  But, it seems like if this option was turned on that it should be possible to build an inheritance graph that would make this easy to do.

Thanks for talking through this!

**Comment by travisdowns** [comment]
I tend to agree with @darlingm that it would be useful to have a variation on this which reported only on classes with an explicit superclass (not the implicit `object` superclass).

I get the justification for the strict check for the "mixin case" but in practice this check cannot be enabled in most codebases which will have a large number of `object`-superclass classes but would very much like the warning about missing `super().__init__` elsewhere. It seems easy in this case to "have your cake and eat it too" with a variation.

**Comment by JanEricNitschke** [comment]
The pylance docs even have the wrong behaviour here: https://github.com/microsoft/pylance-release/blob/main/docs/diagnostics/reportMissingSuperCall.md
